tls/x509_crt.c

```diff
@@ -924,76 +924,58 @@ ttls_x509_crt_parse(TlsX509Crt *chain, unsigned char *buf, size_t buflen)
 	}
 
 	if (buf_format == TTLS_X509_FORMAT_DER)
-		return ttls_x509_crt_parse_der(chain, buf, buflen);
-
-	if (buf_format == TTLS_X509_FORMAT_PEM) {
-		/*
-		 * 1 rather than 0 since the terminating NULL byte
-		 * is counted in.
-		 */
-		while (buflen > 1) {
-			int r;
-			size_t use_len;
-			unsigned char *pem_dec;
+		if ((r = ttls_x509_crt_parse_der(crt, buf, buflen)))
+			goto done;
 
-			/*
-			 * If we get there, we know the string is
-			 * null-terminated.
-			 */
-			r = ttls_pem_read_buffer("-----BEGIN CERTIFICATE-----",
-						 "-----END CERTIFICATE-----",
-						 buf, &use_len);
-			if (r > 0) {
-				/* Was PEM encoded. */
-				pem_dec = buf;
-				buflen -= use_len;
-				buf += use_len;
-			}
-			else if (r != TTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT) {
-				/* PEM header and footer were found. */
-				buflen -= use_len;
-				buf += use_len;
-
-				if (!first_error)
-					first_error = r;
-				total_failed++;
-				continue;
-			}
-			else {
-				break;
-			}
+	if (buf_format != TTLS_X509_FORMAT_PEM)
+		return TTLS_ERR_X509_CERT_UNKNOWN_FORMAT;
```

Comment:
Seems the loop `for ( ; buflen > 1 && crt_len_len; crt_len_len -= TTLS_CERT_LEN_LEN)` is unreachable for DER. If we successfully parsed DER in `ttls_x509_crt_parse_der()` line 927, we just fall to current location, but in this place we will return `TTLS_ERR_X509_CERT_UNKNOWN_FORMAT`

Reply:
Yeah, there is a bug. But we don't need to execute the loop for `TTLS_X509_FORMAT_DER`. In this case we just run `ttls_x509_crt_parse_der()` once and return. But if the format is PEM, then we decode it and parse the encoded DER many times in the loop. Other formats aren't supported, so we return `TTLS_ERR_X509_CERT_UNKNOWN_FORMAT`